### ISSUE TITLE
steps_per_execution autotune

### ISSUE THREAD

**Issue by kvarekamp** [keras-team/keras#16573]
**System information**.
TensorFlow 2.9
Are you willing to contribute it (No)

**Describe the feature and the current behavior/state**.
`steps_per_execution` reduces the number of round trips to the GPU and as such makes it possible to make better use of the GPU without having to resort to larger batch sizes

It would be nice if the number does not need to be specified but can be determined automatically to maximize performance - in the same spirit as tf.data.AUTOTUNE. I have no idea how that could be accomplished though.

**Will this change the current api? How?**
No - it could be a magic constant like AUTOTUNE

**Who will benefit from this feature?**
Everyone

**Comment by tilakrayal** [comment]
Hi @kvarekamp,
Could you elaborate about your Feature. Also, please specify the Use Cases for this feature. Thank you!

**Comment by kvarekamp** [comment]
`model.compile()` has a `steps_per_execution` param that allows you to send say 10 batches to the GPU at once. Rather than tuning this param manually (increasing it until the GPU starts throwing errors) it would be nice if keras can tune it for you and set it to the max number of batches that your GPU can handle (for example by setting steps_per_execution=AUTO).  The use case would be any model running on a GPU with room to spare.

**Comment by rchao** [comment]
Thanks @kvarekamp for the request! We have a autotuning callback that's currently WIP. We are currently a bit short on resources but this remains a high priority.

**Comment by rivershah** [comment]
I would like to better understand why this parameter needs autotuning. Will appreciate if someone can help by looking over these questions:

Is it sufficient to always set `steps_per_execution` to a very high number?
What impact does setting `steps_per_execution` to a high number have on memory, CPU, and device resource utilization?
Are there any concerns about model accuracy when using a very high `steps_per_execution`, or will models with different `steps_per_execution` values always converge to the same metrics? (In contrast, very large batch sizes can negatively impact model performance, as discussed in this [post and paper](https://stats.stackexchange.com/questions/164876/what-is-the-trade-off-between-batch-size-and-number-of-iterations-to-train-a-neu)).
For distributed strategies such as `TPUStrategy`, is there any concern about setting a very large `steps_per_execution`? When do the gradient all-reduces happen across pod devices when using large `steps_per_execution` values? Does the optimizer.apply_gradients behavior change with large `steps_per_execution` values?

**Comment by rchao** [comment]
Also cc'ing @grasskin as Gabriel has been researching into this.

`steps_per_execution` is indeed one of the key parameters that affect the performance (and possibly, to a lesser extent, the model quality) we're trying to understand. The only knowledge we have currently is that a number that's too low would make the training suffer from low throughput. As for the upper bound, we have less information, but in theory the steps per execution can't be so large comparing with steps per epoch that some workers end up being idle because a fraction of the workers have taken up all the steps. A large steps per execution likely also have negative impact when the eviction/preemption of the workers is common. The training step though, including the forward and backward (gradient application) passes, is in theory not affected by the steps per execution.

**Comment by rivershah** [comment]
Setting the `steps_per_execution` parameter to a large value can speed up the training process on toy datasets (device utilization goes up), but it may cause problems when training on larger datasets or using medium-sized TPU pods such as v3-128. In these cases, setting `steps_per_execution` to a large value can lead to nan losses, even without using learning rate scheduling callbacks. This issue can be difficult to debug because it only occurs when training with a large number of devices. It would be helpful to have a tutorial or resource that provides guidance, details / gotchas on the appropriate upper bounds for `steps_per_execution` based on the size of the dataset and the number of replicas and batch sizes being used.

**Comment by innat** [comment]
> Thanks @kvarekamp for the request! We have a autotuning callback that's currently WIP. We are currently a bit short on resources but this remains a high priority.

Hi @rchao, any update on this?

**Comment by grasskin** [comment]
Hi @innat, I'm preparing to release a callback that autotunes `steps_per_execution`. We've been able to see a significant speedup especially in distributed systems. The callback is still going through testing but we're hoping to release it this quarter. I'll update this issue when it's released!

**Comment by feelmyears** [comment]
Hi @grasskin. Any updated timeline on this callback? Would love to put it into action. Thanks

**Comment by grasskin** [comment]
Hi @feelmyyears, we still have to add more information on this in documentation and tutorials, but if you're using the nightly version of Keras you can try it out by setting `compile(..., steps_per_execution='auto')`. Let me know if you run into any issues.

**Comment by grasskin** [comment]
Note this is still highly experimental and final implementation/results might vary.
